### PR TITLE
pretty large refactoring of the Stack Class and surrounding

### PR DESCRIPTION
**Pull request by chregu**
* Add `Stack::getDynamicUriString()`
* Add the `StackUri` and '`UriComponents` classes.
* Add `UriHelper::composeUri(array|UriComponents $components): UriInterface` and `UriHelper::decomposeUri(UriInterface $uri): UriComponents` 
* Add `UriHelper::getSrcSetUrl(string $size, UriInterface $uri, null|string $custom = null)` and `UriHelper::getSrcSetUrlString(string $size, string $uri, null|string $custom = null)`

### CODE REVIEW

File: src/Core/StackUri.php [hunk 1]
```diff
@@ -0,0 +1,151 @@
+<?php
+
+namespace Rokka\Client\Core;
+
+use Psr\Http\Message\UriInterface;
+use Rokka\Client\UriHelper;
+
+/**
+ * Class StackUri.
```
Comment (dbu) [inline src/Core/StackUri.php]: Lets only have the description below. This line adds no information.

File: src/Core/StackUri.php [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        }
+        $config['options'] = $this->getStackOptions();
+
+        return $config;
+    }
+
+    /**
+     * @since 1.2.0
+     *
+     * @param string $options
```
Comment (dbu) [inline src/Core/StackUri.php]: I think we should document the format of options

File: src/Core/UriComponents.php [hunk 3]
```diff
@@ -0,0 +1,189 @@
+<?php
+
+namespace Rokka\Client\Core;
+
+/**
+ * Class UriComponents.
+ *
+ * Abstracts the components away of a rokka uri (has, format, filename, stack, etc)
```
Comment (dbu) [inline src/Core/UriComponents.php]: s/has/hash ?

File: src/UriHelper.php [hunk 4]
```diff
@@ -30,75 +34,222 @@ public static function addOptionsToUriString($url, $options)
...
+        $hashPattern = '(?<hash>[0-9a-f]{6,40})';
+        $filenamePattern = '(?<filename>[A-Za-z\-\0-\9]+)';
+        $formatPattern = '(?<format>.{3,4})';
+        $pathPattern = '(?<hash>-.+-)';
 
-        $optionKeys = array_filter($options, function ($key) {
-            return !($key % 2);
-        }, ARRAY_FILTER_USE_KEY);
+        if (preg_match('#^/'.$stackPattern.'/'.$hashPattern.'/'.$filenamePattern.'\.'.$formatPattern.'$#', $uri->getPath(), $matches)) {
+            // hash with seo-filename
```
Comment (dbu) [inline src/UriHelper.php]: Could you combine these with || ? All conditions dont do anything...

**Review by dbu — COMMENTED** [review]
I like the direction this is going. Would it make sense to use a PSR Uri implementation instead of strings, to need fewer string manipulation?

**Comment by chregu** [comment]
> I like the direction this is going. Would it make sense to use a PSR Uri implementation instead of strings, to need fewer string manipulation?

Do you have an example where that could help us? I tried to use it, whenever possible, but may have missed some parts. and it's mostly about manipulating the "path" part of the URI, where it wouldn't help

**Comment by dbu** [comment]
>> Would it make sense to use a PSR Uri implementation instead of strings, to need fewer string manipulation?
> Do you have an example where that could help us? 

sorry, looking at the places again i realize its about building the paths from the stack information parts. nothing that psr can help with.

**Comment by chregu** [comment]
@dbu Improved the stuff you remarked. Overall documentation could and should be improved, I'll try to come up with something better together with this branch https://github.com/rokka-io/rokka-client-php/tree/templatehelper-from-refactoring-stack (which adds more helper methods for using the client in plugins et al.)

**Review by dbu — APPROVED** [review]
looks good to me. the StackTest is not very complete - did you try using the client a bit with these changes as a plausibility check?

**Comment by chregu** [comment]
I used some variants of this branch already for the drupal and the kirby plugins (which at least create stacks)

**Comment by dbu** [comment]
cheers!

**Pull request merged**